File: cli/flowc.c
#include "flowc.h"
#include "util.h"
#include "cli.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *get_flowc(void) {
    char exe[4096];

    /* 1. Stessa cartella del binario corrente */
    if (get_exe_path(exe, sizeof(exe)) == 0) {
        char *dir   = path_dirname(exe);
        char *flowc = str_fmt("%s" SEP "flowc" EXE_EXT, dir);
        free(dir);
        if (path_exists(flowc)) return flowc;
        free(flowc);
    }

    /* 2. ~/.flow/bin/flowc */
    char *home  = get_flow_home();
    char *flowc = str_fmt("%s" SEP "bin" SEP "flowc" EXE_EXT, home);
    free(home);
    if (path_exists(flowc)) return flowc;
    free(flowc);

    return NULL;
}

int run_flowc(const char *input, const char *outdir, const char *runtime,
              const char *flow_stdlib, int prod, int run_flag, int dev, int fast,
              const char *server_fns) {
    char *flowc = get_flowc();
    if (!flowc) {
        fprintf(stderr, C_RED "✗" C_RESET " flowc non trovato"
                " — esegui prima: flow update\n");
        return 1;
    }

    char srv_fns_arg[4096] = {0};
    if (server_fns && server_fns[0])
        snprintf(srv_fns_arg, sizeof(srv_fns_arg), "--server-fns=%s", server_fns);

    const char *argv[32];
    int i = 0;
    argv[i++] = flowc;
    argv[i++] = input;
    argv[i++] = "--outdir";
    argv[i++] = outdir;
    if (runtime && runtime[0]) {
        argv[i++] = "--runtime";
        argv[i++] = runtime;
    }
    if (flow_stdlib && flow_stdlib[0]) {
        argv[i++] = "--flow";
        argv[i++] = flow_stdlib;
    }
    if (prod)     argv[i++] = "--prod";
    if (run_flag) argv[i++] = "--run";
    if (dev)      argv[i++] = "--dev";
    if (fast)     argv[i++] = "--fast";
    if (srv_fns_arg[0]) argv[i++] = srv_fns_arg;
    argv[i]   = NULL;

    int code = run_cmd(argv);
    free(flowc);
    return code;
}

int run_flowc_route(const char *input, const char *outdir, const char *runtime,
                    const char *flow_stdlib) {
    char *flowc = get_flowc();
    if (!flowc) {
        fprintf(stderr, C_RED "✗" C_RESET " flowc non trovato\n");
        return 1;
    }

    const char *argv[24];
    int i = 0;
    argv[i++] = flowc;
    argv[i++] = input;
    argv[i++] = "--outdir";
    argv[i++] = outdir;
    argv[i++] = "--outlet";
    if (runtime && runtime[0])         { argv[i++] = "--runtime"; argv[i++] = runtime;    }
    if (flow_stdlib && flow_stdlib[0]) { argv[i++] = "--flow";    argv[i++] = flow_stdlib; }
    argv[i] = NULL;

    int code = run_cmd(argv);
    free(flowc);
    return code;
}

int run_flowc_server(const char *input, const char *outdir, const char *runtime,
                     const char *flow_stdlib, int prod) {
    char *flowc = get_flowc();
    if (!flowc) {
        fprintf(stderr, C_RED "✗" C_RESET " flowc non trovato\n");
        return 1;
    }

    const char *argv[24];
    int i = 0;
    argv[i++] = flowc;
    argv[i++] = input;
    argv[i++] = "--outdir";
    argv[i++] = outdir;
    argv[i++] = "--server";
    if (runtime && runtime[0]) { argv[i++] = "--runtime"; argv[i++] = runtime; }
    if (flow_stdlib && flow_stdlib[0]) { argv[i++] = "--flow"; argv[i++] = flow_stdlib; }
    if (prod) argv[i++] = "--prod";
    argv[i] = NULL;

    int code = run_cmd(argv);
    free(flowc);
    return code;
}

File: cli/flowc.ts
// Gestisce il compiler C (flowc) — lo compila al primo avvio e lo cachea
// Il sorgente C è embedded nel binario tramite compiler-bundle.ts

import { join }                      from "node:path"
import { existsSync, mkdirSync,
         writeFileSync }             from "node:fs"
import { COMPILER_BUNDLE }          from "../devtools/compiler-ts-unused/compiler-bundle"

const FLOW_HOME     = join(process.env.USERPROFILE ?? process.env.HOME ?? ".", ".flow")
const COMPILER_DIR  = join(FLOW_HOME, "compiler")
const FLOWC_EXE     = join(COMPILER_DIR, process.platform === "win32" ? "flowc.exe" : "flowc")
const STAMP_FILE    = join(COMPILER_DIR, ".built")

// Estrae i sorgenti C del compiler e compila flowc
async function buildFlowc(): Promise<boolean> {
    mkdirSync(join(COMPILER_DIR, "lexer"),   { recursive: true })
    mkdirSync(join(COMPILER_DIR, "parser"),  { recursive: true })
    mkdirSync(join(COMPILER_DIR, "codegen"), { recursive: true })

    for (const [relPath, content] of Object.entries(COMPILER_BUNDLE)) {
        writeFileSync(join(COMPILER_DIR, relPath), content as string)
    }

    process.stdout.write("  Compilazione flowc (prima volta)...")
    const proc = Bun.spawnSync([
        "clang",
        join(COMPILER_DIR, "main.c"),
        join(COMPILER_DIR, "lexer/lexer.c"),
        join(COMPILER_DIR, "parser/parser.c"),
        join(COMPILER_DIR, "codegen/codegen.c"),
        "-o", FLOWC_EXE,
        "-O2", "-w"
    ])

    if (proc.exitCode === 0) {
        writeFileSync(STAMP_FILE, "1")
        console.log(" ✓")
        return true
    } else {
        console.log(" ✗")
        console.error(new TextDecoder().decode(proc.stderr).split("\n").slice(0, 5).join("\n"))
        return false
    }
}

// Ritorna il path a flowc, compilandolo se necessario
export async function getFlowc(): Promise<string | null> {
    // In sviluppo: usa flowc.exe dalla cartella compiler/
    const devFlowc = join(import.meta.dir, "../compiler/flowc.exe")
    if (existsSync(devFlowc)) return devFlowc

    // Produzione: usa quello cached in ~/.flow/compiler/
    if (existsSync(FLOWC_EXE) && existsSync(STAMP_FILE)) return FLOWC_EXE

    // Prima volta: compila
    const ok = await buildFlowc()
    return ok ? FLOWC_EXE : null
}

// Esegui flowc su un file .flow
export async function runFlowc(
    inputPath: string,
    outDir:    string,
    opts: { run?: boolean; prod?: boolean } = {}
): Promise<boolean> {
    const flowc = await getFlowc()
    if (!flowc) return false

    const args = [flowc, inputPath, "--outdir", outDir]
    if (opts.run)  args.push("--run")
    if (opts.prod) args.push("--prod")

    const proc = Bun.spawnSync(args, { stdio: ["inherit", "inherit", "inherit"] })
    return proc.exitCode === 0
}
